### PR TITLE
Support setting reference types in annotations

### LINKED ISSUE #139: Support for `weak` mock properties
I would like to be able to specify that certain properties should be generated with the `weak` annotation. What would be a good way to add this to Mockolo's capabilities?

### PR DESCRIPTION
**Pull request by tinder-maxwellelliott**
Allows developers to set the reference type of variables in their `@mockable` annotations like so:

`// @mockable(referenceType: someVar = weak)`

Fixes #139

### CODE REVIEW

File: Sources/MockoloFramework/Models/ParsedEntity.swift [hunk 1]
```diff
@@ -100,12 +100,13 @@ final class EntityNodeSubContainer {
...
 
 /// Contains arguments to annotation
-/// e.g. @mockable(module: prefix = Foo; typealias: T = Any; U = String; rx: barStream = PublishSubject; history: bazFunc = true)
+/// e.g. @mockable(module: prefix = Foo; typealias: T = Any; U = String; rx: barStream = PublishSubject; history: bazFunc = true, referenceType: someVar = weak)
 struct AnnotationMetadata {
     var module: String?
     var typeAliases: [String: String]?
     var varTypes: [String: String]?
     var funcsWithArgsHistory: [String]?
+    var referenceTypes: [String: ReferenceType]?
```
Comment (EspressoCup) [inline Sources/MockoloFramework/Models/ParsedEntity.swift:109]: Call this 'modifiers'. The 'reference' makes sense for 'weak' but not for others like 'dynamic'.

File: Sources/MockoloFramework/Models/VariableModel.swift [hunk 2]
```diff
@@ -91,10 +94,19 @@ final class VariableModel: Model {
             return rxVar
         }
 
+        let referenceType: ReferenceType
+        if let referenceTypes = self.referenceTypes,
+           let overrideReferenceType: ReferenceType = referenceTypes[identifier] {
+            referenceType = overrideReferenceType
+        } else {
+            referenceType = .default
```
Comment (EspressoCup) [inline Sources/MockoloFramework/Models/VariableModel.swift:102]: '.none' makes more sense.

File: Sources/MockoloFramework/Templates/VariableTemplate.swift [hunk 3]
```diff
@@ -16,46 +16,60 @@
 
 import Foundation
 
+internal enum ReferenceType: String {
+    case `default` = "default"
```
Comment (EspressoCup) [inline Sources/MockoloFramework/Templates/VariableTemplate.swift:20]: `case none` makes more sense.

File: Sources/MockoloFramework/Models/VariableModel.swift [hunk 4]
```diff
@@ -14,6 +14,7 @@ final class VariableModel: Model {
     var isStatic = false
     var shouldOverride = false
     var overrideTypes: [String: String]?
+    var referenceTypes: [String: ReferenceType]?
```
Comment (EspressoCup) [inline Sources/MockoloFramework/Models/VariableModel.swift:17]: 'modifiers' ditto below

File: Sources/MockoloFramework/Parsers/SwiftSyntaxExtensions.swift [hunk 5]
```diff
@@ -710,8 +711,16 @@ extension Trivia {
                         }
                     }
                 }
-                if argsStr.contains(String.historyColon), let subStr = argsStr.components(separatedBy: String.historyColon).last, !subStr.isEmpty {
```
Comment (EspressoCup) [inline Sources/MockoloFramework/Parsers/SwiftSyntaxExtensions.swift:713]: is this deleted?

File: Tests/TestReferenceTypes/FixtureReferenceTypes.swift [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+            bazFuncHandler(arg, other)
+        }
+
+    }
+}
+
+"""
+
+let referenceTypesWithDynamicAnnotation = """
+/// \(String.mockAnnotation)(referenceType: listener = dynamic)
```
Comment (EspressoCup) [inline Tests/TestReferenceTypes/FixtureReferenceTypes.swift:47]: Add a test case for a dynamic func also. Please verify the generated mocks are valid (try compiling them in source code).

**Review by EspressoCup — COMMENTED** [review]
Thanks for the PR. Left a few comments.

File: Sources/MockoloFramework/Templates/MethodTemplate.swift [hunk 7]
```diff
@@ -118,34 +119,41 @@ extension MethodModel {
                 \(2.tab)set { }
                 """
             }
-            
+
             let overrideStr = isOverride ? "\(String.override) " : ""
+            let modifierTypeString: String
+            switch modifier {
+                case .dynamic, .weak:
+                    modifierTypeString = (overrideStr.isEmpty ? "" :  " ") + "dynamic" + " "
```
Comment (EspressoCup) [inline Sources/MockoloFramework/Templates/MethodTemplate.swift:127]: why hardcode "dynamic" here? Also if no need to look up the overrideStr to prepend an extra space here; it's already added to the overrideStr if non-empty.

File: Sources/MockoloFramework/Templates/VariableTemplate.swift [hunk 8]
```diff
@@ -16,46 +16,60 @@
...
         let privateSetSpace = allowSetCallCount ? "" :  "\(String.privateSet) "
         let setCallCountStmt = "\(underlyingSetCallCount) += 1"
-        
+
+        let modifierTypeStr: String
+        switch modifier {
+        case .none:
+            modifierTypeStr = ""
+        case .dynamic, .weak:
+            modifierTypeStr = (overrideStr.isEmpty ? "" :  " ") + modifier.rawValue + " "
```
Comment (EspressoCup) [inline Sources/MockoloFramework/Templates/VariableTemplate.swift:62]: ditto on the extra space ^

File: Tests/TestReferenceTypes/FixtureReferenceTypes.swift [hunk 9]
```diff
@@ -0,0 +1,130 @@
+import MockoloFramework
+
+let referenceTypesWithWeakAnnotation = """
```
Comment (EspressoCup) [inline Tests/TestReferenceTypes/FixtureReferenceTypes.swift:3]: Change names to modifier--. Also file name s/FixtureReferenceTypes/FixtureModifiers.

File: Tests/TestReferenceTypes/ReferenceTypesTests.swift [hunk 10]
```diff
@@ -0,0 +1,20 @@
+import Foundation
+
+class ReferenceTypesTests: MockoloTestCase {
```
Comment (EspressoCup) [inline Tests/TestReferenceTypes/ReferenceTypesTests.swift:3]: s/ReferenceTypes/Modifiers everywhere and in the file name.

File: Sources/MockoloFramework/Parsers/SwiftSyntaxExtensions.swift [hunk 11]
```diff
@@ -260,111 +260,111 @@ extension ProtocolDeclSyntax: EntityNode {
...
-    
+
     func subContainer(metadata: AnnotationMetadata?, declType: DeclType, path: String?, data: Data?, isProcessed: Bool) -> EntityNodeSubContainer {
         return self.members.members.memberData(with: accessLevel, declType: declType, metadata: nil, processed: isProcessed)
     }
 }
 
 extension VariableDeclSyntax {
-    func models(with acl: String, declType: DeclType, overrides: [String: String]?, processed: Bool) -> [Model] {
+    func models(with acl: String, declType: DeclType, overrides: [String: String]?, modifiers: [String : Modifier]?, processed: Bool) -> [Model] {
```
Comment (EspressoCup) [inline Sources/MockoloFramework/Parsers/SwiftSyntaxExtensions.swift:349]: remove 'modifiers' -- it should be part of 'overrides'

File: Sources/MockoloFramework/Parsers/SwiftSyntaxExtensions.swift [hunk 12]
```diff
@@ -413,17 +415,24 @@ extension SubscriptDeclSyntax {
...
         }
-        
+
         let params = self.signature.input.parameterList.compactMap { $0.model(inInit: false, declType: declType) }
         let genericTypeParams = self.genericParameterClause?.genericParameterList.compactMap { $0.model(inInit: false) } ?? []
         let genericWhereClause = self.genericWhereClause?.description
-        
+
+        let modifier: Modifier
+        if let passedModifiers = modifiers {
```
Comment (EspressoCup) [inline Sources/MockoloFramework/Parsers/SwiftSyntaxExtensions.swift:430]: This should be moved to SwiftSyntaxExtension - 'Trivia.metadata(with annotation: String, in val: String)'

File: Sources/MockoloFramework/Models/VariableModel.swift [hunk 13]
```diff
@@ -91,10 +94,19 @@ final class VariableModel: Model {
             return rxVar
         }
 
+        let modifier: Modifier
+        if let modifiers = self.modifiers,
```
Comment (EspressoCup) [inline Sources/MockoloFramework/Models/VariableModel.swift:98]: This should be moved to SwiftSyntaxExtension - 'Trivia.metadata(with annotation: String, in val: String)'

File: Sources/MockoloFramework/Models/VariableModel.swift [hunk 14]
```diff
@@ -52,11 +54,12 @@ final class VariableModel: Model {
         self.canBeInitParam = canBeInitParam
         self.processed = processed
         self.overrideTypes = overrideTypes
+        self.modifiers = modifiers
```
Comment (EspressoCup) [inline Sources/MockoloFramework/Models/VariableModel.swift:57]: remove - this should be part of 'overrideTypes'

File: Sources/MockoloFramework/Parsers/SwiftSyntaxExtensions.swift [hunk 15]
```diff
@@ -132,32 +132,32 @@ extension MemberDeclListItemSyntax {
...
         return modifiers?.acl ?? ""
     }
-    
+
     func transformToModel(with encloserAcl: String, declType: DeclType, metadata: AnnotationMetadata?, processed: Bool) -> (Model, String?, Bool)? {
         if let varMember = self.decl.as(VariableDeclSyntax.self) {
             if validateMember(varMember.modifiers, declType, processed: processed) {
                 let acl = memberAcl(varMember.modifiers, encloserAcl, declType)
-                if let item = varMember.models(with: acl, declType: declType, overrides: metadata?.varTypes, processed: processed).first {
+                if let item = varMember.models(with: acl, declType: declType, overrides: metadata?.varTypes, modifiers: metadata?.modifiers, processed: processed).first {
```
Comment (EspressoCup) [inline Sources/MockoloFramework/Parsers/SwiftSyntaxExtensions.swift:153]: remove modifiers param here - it should be part of 'overrides'

File: Sources/MockoloFramework/Parsers/SwiftSyntaxExtensions.swift [hunk 16]
```diff
@@ -476,13 +486,14 @@ extension InitializerDeclSyntax {
                            params: params,
                            throwsOrRethrows: self.throwsOrRethrowsKeyword?.text ?? "",
                            isStatic: false,
+                           modifier: .none,
```
Comment (EspressoCup) [inline Sources/MockoloFramework/Parsers/SwiftSyntaxExtensions.swift:489]: remove this

File: Sources/MockoloFramework/Templates/VariableTemplate.swift [hunk 17]
```diff
@@ -16,46 +16,60 @@
 
 import Foundation
 
+internal enum Modifier: String {
+    case none = "none"
```
Comment (EspressoCup) [inline Sources/MockoloFramework/Templates/VariableTemplate.swift:20]: case none = ""

File: Sources/MockoloFramework/Parsers/SwiftSyntaxExtensions.swift (hunk 11 above)
Comment (EspressoCup) [inline Sources/MockoloFramework/Parsers/SwiftSyntaxExtensions.swift:349]: rename --> customModifiers to differentiate from built-in modifiers.

File: Sources/MockoloFramework/Parsers/SwiftSyntaxExtensions.swift [hunk 18]
```diff
@@ -413,17 +415,24 @@ extension SubscriptDeclSyntax {
 }
 
 extension FunctionDeclSyntax {
-    
-    func model(with acl: String, declType: DeclType, funcsWithArgsHistory: [String]?, processed: Bool) -> Model {
+
+    func model(with acl: String, declType: DeclType, funcsWithArgsHistory: [String]?, modifiers: [String : Modifier]?, processed: Bool) -> Model {
```
Comment (EspressoCup) [inline Sources/MockoloFramework/Parsers/SwiftSyntaxExtensions.swift:419]: rename --> customModifiers to differentiate from built-in modifiers.

File: Sources/MockoloFramework/Templates/VariableTemplate.swift [hunk 19]
```diff
@@ -16,46 +16,60 @@
 
 import Foundation
 
+internal enum Modifier: String {
```
Comment (EspressoCup) [inline Sources/MockoloFramework/Templates/VariableTemplate.swift:19]: Remove 'internal'. Move this to ParsedEntity.swift.

File: Sources/MockoloFramework/Models/VariableModel.swift [hunk 20]
```diff
@@ -41,6 +42,7 @@ final class VariableModel: Model {
          offset: Int64,
          length: Int64,
          overrideTypes: [String: String]?,
+         modifiers: [String: Modifier]?,
```
Comment (EspressoCup) [inline Sources/MockoloFramework/Models/VariableModel.swift:45]: rename --> customModifiers

File: Sources/MockoloFramework/Templates/MethodTemplate.swift [hunk 21]
```diff
@@ -123,10 +123,10 @@ extension MethodModel {
             let overrideStr = isOverride ? "\(String.override) " : ""
             let modifierTypeString: String
             switch modifier {
-                case .dynamic, .weak:
-                    modifierTypeString = (overrideStr.isEmpty ? "" :  " ") + "dynamic" + " "
-                default:
+                case .none:
                     modifierTypeString = ""
+                default:
```
Comment (elsh) [inline Sources/MockoloFramework/Templates/MethodTemplate.swift:128]: a method can't be "weak"

File: Sources/MockoloFramework/Templates/MethodTemplate.swift [hunk 22]
```diff
@@ -123,10 +123,10 @@ extension MethodModel {
...
+                    modifierTypeString = modifier.rawValue + " "
```
Comment (EspressoCup) [inline Sources/MockoloFramework/Templates/MethodTemplate.swift:129]: a method can't be "weak"

**Review by EspressoCup — APPROVED** [review]
Thanks for the updates, lgtm.

**Pull request merged**